File: test_wasi.js
#!/usr/bin/env node
/**
 * Test script for the execute_query WASI module.
 * Node.js version of test_wasi.py
 */

const { WASI } = require('wasi');
const { readFile } = require('fs/promises');
const { argv, env } = require('process');

class WasiExecuteQuery {
    /**
     * Wrapper class for execute_query WASI module that supports multiple calls.
     */
    constructor(verbose = true) {
        this.verbose = verbose;
        this.instance = null;
        this.memory = null;
        this.wasm_malloc = null;
        this.wasm_free = null;
        this.execute_query_main = null;
        this._initialized = false;
    }

    async load(wasmPath) {
        if (this.verbose) {
            console.log(`Loading module: ${wasmPath}`);
        }

        const wasi = new WASI({
            version: 'preview1',
            args: ['execute_query'],
            env: {},
        });

        const wasmBuffer = await readFile(wasmPath);
        
        if (this.verbose) {
            console.log('Compiling module...');
        }
        
        const module = await WebAssembly.compile(wasmBuffer);
        
        if (this.verbose) {
            console.log('Instantiating module...');
        }

        this.instance = await WebAssembly.instantiate(module, {
            wasi_snapshot_preview1: wasi.wasiImport,
        });

        // Get exports
        this.memory = this.instance.exports.memory;
        this.wasm_malloc = this.instance.exports.wasm_malloc;
        this.wasm_free = this.instance.exports.wasm_free;
        this.execute_query_main = this.instance.exports.execute_query_main;
        const _initialize = this.instance.exports._initialize;

        if (this.verbose) {
            console.log(`Exports found: memory, wasm_malloc, wasm_free, execute_query_main, _initialize=${_initialize !== undefined}`);
        }

        // Initialize WASI (this calls _initialize internally for reactor modules)
        // Note: wasi.initialize() already handles _initialize for reactor modules
        wasi.initialize(this.instance);

        this._initialized = true;
    }

    /**
     * Execute query with given arguments.
     * @param {string[]} args - List of arguments (without 'execute_query' prefix)
     * @param {boolean|null} verbose - Override instance verbose setting
     * @returns {number} Exit code from execute_query_main
     */
    execute(args, verbose = null) {
        if (verbose === null) {
            verbose = this.verbose;
        }

        // Prepare arguments
        const fullArgs = ['execute_query', ...args];

        if (verbose) {
            console.log(`Arguments: ${JSON.stringify(fullArgs)}`);
        }

        // Get memory view
        const memData = new Uint8Array(this.memory.buffer);
        const memView = new DataView(this.memory.buffer);

        // Allocate memory for argv array
        const argc = fullArgs.length;
        const argvSize = argc * 4; // 4 bytes per pointer (32-bit WASM)
        const argvPtr = this.wasm_malloc(argvSize);

        if (verbose) {
            console.log(`Allocated argv at ${argvPtr}`);
        }

        // Allocate and copy each argument string
        const argPtrs = [];
        const encoder = new TextEncoder();

        for (let i = 0; i < fullArgs.length; i++) {
            const arg = fullArgs[i];
            const argBytes = encoder.encode(arg + '\0');
            const argPtr = this.wasm_malloc(argBytes.length);

            if (verbose) {
                console.log(`  arg[${i}] = '${arg}' at ${argPtr}`);
            }

            // Copy string bytes to memory
            memData.set(argBytes, argPtr);
            argPtrs.push(argPtr);
        }

        // Write argv pointers to the argv array (little-endian 32-bit)
        for (let i = 0; i < argPtrs.length; i++) {
            memView.setUint32(argvPtr + i * 4, argPtrs[i], true);
        }

        if (verbose) {
            console.log(`Calling execute_query_main(${argc}, ${argvPtr})...`);
        }

        let result = null;
        try {
            result = this.execute_query_main(argc, argvPtr);
            if (verbose) {
                console.log(`Result: ${result}`);
            }
        } catch (e) {
            if (e.message && e.message.includes('exit')) {
                // Extract exit code from error
                const match = e.message.match(/exit\((\d+)\)/);
                if (match) {
                    result = parseInt(match[1]);
                    if (verbose) {
                        console.log(`Exit status: ${result}`);
                    }
                } else {
                    console.error(`Error: ${e.message}`);
                    throw e;
                }
            } else {
                console.error(`Error: ${e}`);
                throw e;
            }
        }

        // Free memory
        for (const ptr of argPtrs) {
            this.wasm_free(ptr);
        }
        this.wasm_free(argvPtr);

        return result;
    }
}

/**
 * Load the WASI module and execute a query (legacy function).
 */
async function runExecuteQuery(wasmPath, queryArgs = []) {
    const executor = new WasiExecuteQuery(true);
    await executor.load(wasmPath);
    return executor.execute(queryArgs);
}

/**
 * Run benchmark with multiple queries.
 */
async function runBenchmark(wasmPath, queryArgs, iterations = 10) {
    console.log('='.repeat(60));
    console.log('BENCHMARK MODE');
    console.log('='.repeat(60));

    // Load module once
    console.log('\n[1] Loading module...');
    const loadStart = performance.now();
    const executor = new WasiExecuteQuery(false);
    await executor.load(wasmPath);
    const loadTime = (performance.now() - loadStart) / 1000;
    console.log(`    Module load time: ${loadTime.toFixed(3)}s`);

    // Warm-up run
    console.log('\n[2] Warm-up run...');
    executor.execute(queryArgs, false);

    // Benchmark runs
    console.log(`\n[3] Running ${iterations} iterations...`);
    const times = [];

    for (let i = 0; i < iterations; i++) {
        const start = performance.now();
        const result = executor.execute(queryArgs, false);
        const elapsed = (performance.now() - start) / 1000;
        times.push(elapsed);
        console.log(`    Iteration ${i + 1}: ${(elapsed * 1000).toFixed(2)}ms (result=${result})`);
    }

    // Statistics
    const avgTime = times.reduce((a, b) => a + b, 0) / times.length;
    const minTime = Math.min(...times);
    const maxTime = Math.max(...times);

    console.log('\n' + '='.repeat(60));
    console.log('RESULTS');
    console.log('='.repeat(60));
    console.log(`  Module load time: ${loadTime.toFixed(3)}s`);
    console.log(`  Average execution: ${(avgTime * 1000).toFixed(2)}ms`);
    console.log(`  Min execution: ${(minTime * 1000).toFixed(2)}ms`);
    console.log(`  Max execution: ${(maxTime * 1000).toFixed(2)}ms`);
    console.log(`  Queries per second: ${(1 / avgTime).toFixed(1)}`);
    console.log('='.repeat(60));
}

async function main() {
    const wasmPath = '/mnt/shared/workspaces/opensource/zetasql3/bazel-bin/zetasql/tools/execute_query/execute_query_wasi';
    
    const args = process.argv.slice(2);

    if (args.length > 0 && args[0] === '--benchmark') {
        // Benchmark mode: --benchmark <iterations> <query>
        let iterations = 10;
        let queryArgs = args.slice(1);

        if (queryArgs.length > 0 && /^\d+$/.test(queryArgs[0])) {
            iterations = parseInt(queryArgs[0]);
            queryArgs = queryArgs.slice(1);
        }

        if (queryArgs.length === 0) {
            queryArgs = ['SELECT 1 + 2'];
        }

        await runBenchmark(wasmPath, queryArgs, iterations);
    } else {
        // Normal mode
        await runExecuteQuery(wasmPath, args);
    }
}

main().catch(console.error);

File: test_wasi.py
#!/usr/bin/env python3
"""Test script for the execute_query WASI module."""

import wasmtime
import sys

def run_execute_query(wasm_path: str, query: str = None):
    """Load the WASI module and execute a query."""

    # Configure the engine with threads support
    config = wasmtime.Config()
    config.wasm_threads = True
    
    engine = wasmtime.Engine(config)
    store = wasmtime.Store(engine)
    linker = wasmtime.Linker(engine)
    
    # Add WASI imports
    wasi_config = wasmtime.WasiConfig()
    wasi_config.inherit_stdout()
    wasi_config.inherit_stderr()
    store.set_wasi(wasi_config)
    linker.define_wasi()
    
    # Load the module
    print(f"Loading module: {wasm_path}")
    module = wasmtime.Module.from_file(engine, wasm_path)
    
    # Instantiate
    print("Instantiating module...")
    instance = linker.instantiate(store, module)

    # Get exports
    memory = instance.exports(store)["memory"]
    wasm_malloc = instance.exports(store)["wasm_malloc"]
    wasm_free = instance.exports(store)["wasm_free"]
    execute_query_main = instance.exports(store)["execute_query_main"]
    initialize = instance.exports(store).get("_initialize")
    
    print(f"Exports found: memory, wasm_malloc, wasm_free, execute_query_main, _initialize={initialize is not None}")
    
    # Call _initialize if present (reactor mode)
    if initialize:
        print("Calling _initialize...")
        initialize(store)
        print("_initialize completed")

    # Prepare arguments
    args = ["execute_query"] + query
    
    print(f"Arguments: {args}")
    
    # Allocate memory for argv array
    argc = len(args)
    argv_size = argc * 4  # 4 bytes per pointer (32-bit WASM)
    argv_ptr = wasm_malloc(store, argv_size)
    print(f"Allocated argv at {argv_ptr}")
    
    # Allocate and copy each argument string
    arg_ptrs = []
    for i, arg in enumerate(args):
        arg_bytes = arg.encode('utf-8') + b'\0'
        arg_ptr = wasm_malloc(store, len(arg_bytes))
        print(f"  arg[{i}] = '{arg}' at {arg_ptr}")
        
        # Copy string to WASM memory (handle both shared and regular memory)
        if hasattr(memory, 'data_ptr'):
            try:
                # Try with store argument first (regular memory)
                mem_data = memory.data_ptr(store)
            except TypeError:
                # Shared memory doesn't need store
                mem_data = memory.data_ptr()
        
        for j, b in enumerate(arg_bytes):
            mem_data[arg_ptr + j] = b
        
        arg_ptrs.append(arg_ptr)
    
    # Write argv pointers to the argv array
    if hasattr(memory, 'data_ptr'):
        try:
            mem_data = memory.data_ptr(store)
        except TypeError:
            mem_data = memory.data_ptr()
    for i, ptr in enumerate(arg_ptrs):
        # Write 32-bit pointer (little endian)
        offset = argv_ptr + i * 4
        mem_data[offset] = ptr & 0xFF
        mem_data[offset + 1] = (ptr >> 8) & 0xFF
        mem_data[offset + 2] = (ptr >> 16) & 0xFF
        mem_data[offset + 3] = (ptr >> 24) & 0xFF
    
    print(f"Calling execute_query_main({argc}, {argv_ptr})...")
    result = None
    try:
        result = execute_query_main(store, argc, argv_ptr)
        print(f"Result: {result}")
    except wasmtime.ExitTrap as e:
        print(f"Exit status: {e.code}")
        result = e.code
    except Exception as e:
        print(f"Error: {e}")
    
    # Free memory
    for ptr in arg_ptrs:
        wasm_free(store, ptr)
    wasm_free(store, argv_ptr)
    
    return result

if __name__ == "__main__":
    wasm_path = "/mnt/shared/workspaces/opensource/zetasql3/bazel-bin/zetasql/tools/execute_query/execute_query_wasi"
    
    run_execute_query(wasm_path, sys.argv[1:])
